File: RideShareApp/RideShareApp/Driver.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace RideShareApp
{
    public class Driver : User
    {
        public string VehicleType { get; set; }
        public string Location { get; set; }
        public bool IsAvailable { get; set; } = true;

        public void AcceptRide(Trip trip)
        {
            IsAvailable = false;
            trip.Status = TripStatus.Accepted;
            Console.WriteLine($"{Name} accepted the ride.");
        }

        public void StartTrip(Trip trip)
        {
            trip.Status = TripStatus.InProgress;
            Console.WriteLine($"{Name} started the trip.");
        }

        public void CompleteTrip(Trip trip)
        {
            IsAvailable = true;
            trip.Status = TripStatus.Completed;
            Console.WriteLine($"{Name} completed the trip.");
        }

        public void RateRider(Rider rider, double rating)
        {
            rider.Rating = (rider.Rating + rating) / 2;
            Console.WriteLine($"{Name} rated rider {rider.Name} with {rating} stars");
        }
    }
}

File: RideSharingApp__Strategy_Pattern/src/Driver.java
public class Driver extends User {
    private String vehicleType;
    private boolean isAvailable;

    public Driver(int id, String name, String location, String vehicleType) {
        super(id, name, location);
        this.vehicleType = vehicleType;
        this.isAvailable = true;
    }

    public String getVehicleType() {
        return vehicleType;
    }

    public boolean isAvailable() {
        return isAvailable;
    }

    public void acceptRide(Trip trip) {
        if (isAvailable) {
            trip.assignDriver(this);
            System.out.println(name + " has accepted the ride.");
        }
    }

    public void startTrip(Trip trip) {
        trip.startTrip();
    }

    public void completeTrip(Trip trip) {
        trip.completeTrip();
    }

    public void updateRating(double rating) {
        this.rating = (this.rating + rating) / 2; // Assuming `rating` is inherited from User
    }

    public void rateRider(Rider rider, double rating) {
        System.out.println(name + " rated " + rider.getName() + " with " + rating + " stars.");
    }

    @Override
    public void displayInfo() {
        System.out.println("Driver: " + name + ", Vehicle Type: " + vehicleType + ", Location: " + location + ", Rating: " + rating);
    }
}

